cmd/neofs-node/policy.go

```diff
@@ -32,58 +119,106 @@ func (x *containerNodes) forEachContainerNodePublicKeyInLastTwoEpochs(cnrID cid.
+	caches, err := x.updateAndGetCachesForCurrentEpoch(curEpoch)
 	if err != nil {
-		return fmt.Errorf("read container by ID: %w", err)
+		return err
 	}
 
-	networkMap, err := x.network.GetNetMapByEpoch(epoch)
-	if err != nil {
-		return fmt.Errorf("read network map at epoch #%d: %w", epoch, err)
-	}
-	// TODO(#2692): node sets remain unchanged for fixed container and network map,
-	//  so recently calculated results worth caching
-	ns, err := networkMap.ContainerNodes(cnr.Value.PlacementPolicy(), cnrID)
-	if err != nil {
-		return fmt.Errorf("apply container's storage policy to the network map at epoch #%d: %w", epoch, err)
-	}
+	cnrCtx := containerPolicyContext{id: cnrID, containers: x.containers, network: x.network}
 
-	for i := range ns {
-		for j := range ns[i] {
-			if !f(ns[i][j]) {
-				return nil
+	resCur, err := cnrCtx.applyAtEpoch(curEpoch, caches.cur)
+	if err != nil {
+		return fmt.Errorf("select container nodes for current epoch #%d: %w", curEpoch, err)
+	} else if resCur.err == nil { // error case handled below
+		for i := range resCur.nodeSets {
+			for j := range resCur.nodeSets[i] {
+				if !f(resCur.nodeSets[i][j]) {
+					return nil
+				}
 			}
 		}
 	}
 
-	if !withPrevEpoch || epoch == 0 {
+	if !withPrevEpoch || curEpoch == 0 {
+		if resCur.err != nil {
+			return fmt.Errorf("select container nodes for current epoch #%d: %w", curEpoch, resCur.err)
+		}
 		return nil
 	}
 
-	epoch--
-
-	networkMap, err = x.network.GetNetMapByEpoch(epoch)
+	resPrev, err := cnrCtx.applyAtEpoch(curEpoch-1, caches.prev)
 	if err != nil {
-		return fmt.Errorf("read network map at epoch #%d: %w", epoch, err)
+		if resCur.err != nil {
+			return fmt.Errorf("select container nodes for both epochs: (current#%d) %w; (previous#%d) %w",
+				curEpoch, resCur.err, curEpoch-1, err)
+		}
+		return fmt.Errorf("select container nodes for previous epoch #%d: %w", curEpoch-1, err)
+	} else if resPrev.err == nil { // error case handled below
+		for i := range resPrev.nodeSets {
+			for j := range resPrev.nodeSets[i] {
+				if !f(resPrev.nodeSets[i][j]) {
+					return nil
+				}
+			}
+		}
 	}
 
-	ns, err = networkMap.ContainerNodes(cnr.Value.PlacementPolicy(), cnrID)
-	if err != nil {
-		return fmt.Errorf("apply container's storage policy to the network map at epoch #%d: %w", epoch, err)
+	if resCur.err != nil {
+		if resPrev.err != nil {
+			return fmt.Errorf("select container nodes for both epochs: (current#%d) %w; (previous#%d) %w",
+				curEpoch, resCur.err, curEpoch-1, resPrev.err)
+		}
+		return fmt.Errorf("select container nodes for current epoch #%d: %w", curEpoch, resCur.err)
+	} else if resPrev.err != nil {
+		return fmt.Errorf("select container nodes for previous epoch #%d: %w", curEpoch-1, resPrev.err)
 	}
+	return nil
+}
 
-	for i := range ns {
-		for j := range ns[i] {
-			if !f(ns[i][j]) {
-				return nil
-			}
+// preserves context of storage policy processing for the particular container.
+type containerPolicyContext struct {
```

Comment:
it has only one method, do we need it? make more complex, not helps imo

Reply:
can drop, worth it?